bam_stat.c

```diff
@@ -97,15 +97,130 @@ static void usage_exit(FILE *fp, int exit_status)
+    printf("%lld + %lld singletons (%s : %s)\n", s->n_sgltn[0], s->n_sgltn[1], percent(b0, s->n_sgltn[0], s->n_pair_all[0]), percent(b1, s->n_sgltn[1], s->n_pair_all[1]));
+    printf("%lld + %lld with mate mapped to a different chr\n", s->n_diffchr[0], s->n_diffchr[1]);
+    printf("%lld + %lld with mate mapped to a different chr (mapQ>=5)\n", s->n_diffhigh[0], s->n_diffhigh[1]);
+}
+
+static void out_fmt_json(bam_flagstat_t *s) {
+    char b0[16], b1[16];
+    char *str = "{ \"QC-passed reads\": {\
+                 \"total\": \"%lld\", \
+                 \"secondary\": \"%lld\", \
+                 \"supplementary\": \"%lld\", \
+                 \"duplicates\": \"%lld\", \
+                 \"mapped\": \"%lld (%s)\", \
+                 \"paired in sequencing\": \"%lld\", \
+                 \"read1\": \"%lld\", \
+                 \"read2\": \"%lld\", \
+                 \"properly paired\": \"%lld\", \
+                 \"with itself and mate mapped\": \"%lld\", \
+                 \"singletons\": \"%lld (%s)\", \
+                 \"with mate mapped to a different chr\": \"%lld\", \
+                 \"with mate mapped to a different chr (mapQ >= 5)\": \"%lld\" }, \
+              \"QC-failed reads\": {\
+                 \"total\": \"%lld\", \
+                 \"secondary\": \"%lld\", \
+                 \"supplementary\": \"%lld\", \
+                 \"duplicates\": \"%lld\", \
+                 \"mapped\": \"%lld (%s)\", \
+                 \"paired in sequencing\": \"%lld\", \
+                 \"read1\": \"%lld\", \
+                 \"read2\": \"%lld\", \
+                 \"properly paired\": \"%lld\", \
+                 \"with itself and mate mapped\": \"%lld\", \
+                 \"singletons\": \"%lld (%s)\", \
+                 \"with mate mapped to a different chr\": \"%lld\", \
+                 \"with mate mapped to a different chr (mapQ >= 5)\": \"%lld\" } \
+              }";
+
+    printf(str,
+        s->n_reads[0],
+        s->n_secondary[0],
+        s->n_supp[0],
+        s->n_dup[0],
+        s->n_mapped[0], percent(b0, s->n_mapped[0], s->n_reads[0]),
+        s->n_pair_all[0],
+        s->n_read1[0],
+        s->n_read2[0],
+        s->n_pair_good[0],
+        s->n_pair_map[0],
+        s->n_sgltn[0], percent(b0, s->n_sgltn[0], s->n_pair_all[0]),
+        s->n_diffchr[0],
+        s->n_diffhigh[0],
+        s->n_reads[1],
+        s->n_secondary[1],
+        s->n_supp[1],
+        s->n_dup[1],
+        s->n_mapped[1], percent(b1, s->n_mapped[1], s->n_reads[1]),
+        s->n_pair_all[1],
+        s->n_read1[1],
+        s->n_read2[1],
+        s->n_pair_good[1],
+        s->n_pair_map[1],
+        s->n_sgltn[1], percent(b1, s->n_sgltn[1], s->n_pair_all[1]),
+        s->n_diffchr[1],
+        s->n_diffhigh[1]
+    );
+}
+
+static void out_fmt_tsv(bam_flagstat_t *s) {
+    char b0[16], b1[16];
+    printf("%lld \t %lld \t total (QC-passed reads + QC-failed reads)\n", s->n_reads[0], s->n_reads[1]);
```

Comment:
I also just noticed, this should be `"%lld\t%lld\ttotal (QC-passed reads + QC-failed reads)\n"`, as currently they're strictly space-tab-space separated values.

Reply:
Thanks for pointing out.